### PR TITLE
php store cookies in cookiejar without calling curl_close

### PR DESCRIPTION
**Pull request by frosty00**

### CODE REVIEW

File: php/base/Exchange.php [hunk 1]
```diff
@@ -1346,6 +1346,14 @@ function ($curl, $header) use (&$response_headers, &$http_status_text) {
         $curl_error = curl_error($this->curl);
         $http_status_code = curl_getinfo($this->curl, CURLINFO_HTTP_CODE);
 
+        // Write cookies here but do not close the handle
+        $cookie_file = static::safe_value($this->curl_options, CURLOPT_COOKIEJAR);
+        if ($cookie_file) {
+            $cookies = implode("\n", curl_getinfo($this->curl, CURLINFO_COOKIELIST)) . "\n";
+            $file = fopen($cookie_file, 'w');
```
Comment (kroitor) [inline php/base/Exchange.php]: We're not really supposed to do file io or disk io inside the lib... Instead, I would suggest to merge the options that are necessary to handle the cookie file by means of curl – with the other options we're setting, and without resetting the handle (only enable this if the corresponding controlling properties are set on the exchange instance). And let the user handle file io on the user side. Lmk what you think.

Comment (frosty00) [reply php/base/Exchange.php]: curl does file io to the cookie_jar when curl_close is called if `CURLOPT_COOKIEJAR` is set. I just replicated their design.

Comment (frosty00) [reply php/base/Exchange.php]: `CURLOPT_COOKIEFILE` is the file it reads the cookies from and `CURLOPT_COOKIEJAR` is where it stores the cookies too upon the closing of the handle.

Comment (kroitor) [reply php/base/Exchange.php]: > curl does file io to the cookie_jar when curl_close is called if CURLOPT_COOKIEJAR is set. I just replicated their design.

Yes, but why won't we just leave it for CURL to do that, instead of actually adding fopen to the base code?

Comment (frosty00) [reply php/base/Exchange.php]: because we don't want to close the handle

Comment (frosty00) [reply php/base/Exchange.php]: there is no `curl_save_cookies` function just `curl_close` afaik

Comment (kroitor) [reply php/base/Exchange.php]: > because we don't want to close the handle

Do we have to?

What about these? ↓

- https://www.php.net/manual/en/function.curl-setopt.php#114475
- https://www.php.net/manual/en/function.curl-setopt.php#118967
- https://www.php.net/manual/en/function.curl-setopt.php#91510

File: php/base/Exchange.php [hunk 2]
```diff
@@ -1346,6 +1346,9 @@ function ($curl, $header) use (&$response_headers, &$http_status_text) {
         $curl_error = curl_error($this->curl);
         $http_status_code = curl_getinfo($this->curl, CURLINFO_HTTP_CODE);
 
+        // Write cookies here but do not close the handle
+        $curl_copy = curl_copy_handle($this->curl);
+        curl_close($curl_copy);
```
Comment (kroitor) [inline php/base/Exchange.php]: Closing a copy will not dump them, unfortunately. We need to close the original handle. However, maybe a better solution would be to try in-mem cookies without actually closing the handle. 

Apart from that we only need to do this if the user has set a corresponding property on the exchange or in curl_options, otherwise this becomes unconditional and gets executed upon every request, which is not what most users would want.

File: php/base/Exchange.php (hunk 1 above)
Comment (frosty00) [reply php/base/Exchange.php]: We don't close the handle in the current version.

Comment (frosty00) [reply php/base/Exchange.php]: `curl_close($this-curl)` would throw an error

Comment (kroitor) [reply php/base/Exchange.php]: What kind of error?

File: php/base/Exchange.php (hunk 2 above)
Comment (frosty00) [reply php/base/Exchange.php]: > However, maybe a better solution would be to try in-mem cookies without actually closing the handle.

wdym? my original solution was functioning and had the same behavior as curl... also by doing this we introduce more `last_x_thing` that we are trying to avoid.

Comment (kroitor) [reply php/base/Exchange.php]: I mean this particular type of solution, for example:

- https://www.php.net/manual/en/function.curl-setopt.php#118967

> also by doing this we introduce more last_x_thing that we are trying to avoid.

It's all sync in PHP anyways, so, if the user wants a cookie – something like `last_x_thing` is the only way, it is always a `last_x_thing`-type solution, even with the files. If you think logically, disk io is a `last_x_thing`-type solution. So, it's the same, because it does not matter where you store your `last_x_thing` – in the memory (preferred) or on disk (unwanted, because disk io is slooooow).

I'd suggest that we try in-memory cookies, but it has to be tested, can't skip the checks.

Comment (frosty00) [reply php/base/Exchange.php]: in memory would definitely work, i have tested it. should it be called `last_response_cookies`?

Comment (kroitor) [reply php/base/Exchange.php]: At least, it seems to do the thing without actually messing with the curl handle.

Comment (kroitor) [reply php/base/Exchange.php]: > should it be called last_response_cookies?

No, we should not attempt to parse that and should not attempt to store that. What we should do is:

1. if the user has set a corresponding option in curl_options – enable in-mem cookies... if that can be done by using the curl_options alone without adding any other code – we should just let the user know of how to enable it and close the issue #5914, without adding any code.
2. if that requires additional handling inside the lib – it should only be done, if the user has requested this, otherwise don't waste cycles.
3. after a call – don't parse or save the cookies – let the user (and curl) handle that if necessary, by using the exposed curl handle.

In other words, we want to transparently propagate that option to curl, without messing around with CURL's internals and without actually touching those cookies at all. Does that make sense?

Comment (frosty00) [reply php/base/Exchange.php]: since we call `curl_reset` we lose all the information about the cookies. curl already has in-mem cookies by default but they get deleted after a call to `curl_reset`

File: php/base/Exchange.php [hunk 3]
```diff
@@ -1346,6 +1347,15 @@ function ($curl, $header) use (&$response_headers, &$http_status_text) {
         $curl_error = curl_error($this->curl);
         $http_status_code = curl_getinfo($this->curl, CURLINFO_HTTP_CODE);
 
+        // Save cookies in mem here but do not close the handle
+        $cookie_file = static::safe_value($this->curl_options, CURLOPT_COOKIEJAR);
+        $this->last_response_cookies = curl_getinfo($this->curl, CURLINFO_COOKIELIST);
```
Comment (frosty00) [inline php/base/Exchange.php]: here we can save it in-mem (the fopen code is only run if the user specifies `CURLOPT_COOKIEJAR`)

File: php/base/Exchange.php (hunk 2 above)
Comment (kroitor) [reply php/base/Exchange.php]: > since we call curl_reset we lose all the information about the cookies. curl already has in-mem cookies by default but they get deleted after a call to curl_reset

Exactly, and this is why we have steps 2 and 3 above. Can we avoid the reset after the call completely or do it before the next call (in the beginning of next fetch, not after this fetch)?

Comment (frosty00) [reply php/base/Exchange.php]: that is one of the possible solutions, however it means if the user changed the `$this->curl` object inbetween calls it may not work as expected.

Comment (frosty00) [reply php/base/Exchange.php]: i guess it doesn't matter since we override most of the options in `fetch` anyway. I think this is the best solution.

Comment (kroitor) [reply php/base/Exchange.php]: > that is one of the possible solutions, however it means if the user changed the $this->curl object inbetween calls it may not work as expected.

Well if someone does this, i guess they should know what they're doing ) I mean, overriding the curl handle itself is a bit beyond the scope of the cookie problem... So, I would suggest to keep it as plain and simple as it can possibly be – we already expose everything that we need to let the user enable those cookies, and then the only thing we need to adjust is the time of the reset call? Or maybe even make the reset optional?... What do you think?

Comment (frosty00) [reply php/base/Exchange.php]: I would keep the reset there for sanity reasons and just move it to the top of `fetch`

Comment (kroitor) [reply php/base/Exchange.php]: @frosty00 yep, i think that should work well.

Comment (frosty00) [reply php/base/Exchange.php]: users are already able to specify curl options via the `curl_options` so not a big deal.
